### PR TITLE
feat: add waiting for http response

### LINKED ISSUE #142: Allow to wait for host to be ready
In many cases, more so when in a proxy, we need to wait and block until the proxy (using the WAF) is ready to accept requests.

The way we solved this in coraza-proxy-wasm is that we have a dedicated instruction running a curl and waiting for the host to be available. This is handy but not ideal as we have another piece of complexity for such a simple problem. Also in corporate envs, pulling an image isn't trivial and that is a friction.

I wish go-ftw allows to pass in a URL and it would wait until that URL returns 200.

### PR DESCRIPTION
**Pull request by fzipi**
Adds:
```
      --wait-for-connection-timeout duration   Http connection timeout, The timeout includes connection time, any redirects, and reading the response body. (default 3s)
      --wait-for-expect-body-json string       Expect response body JSON pattern.
      --wait-for-expect-body-regex string      Expect response body pattern.
      --wait-for-expect-body-xpath string      Expect response body XPath pattern.
      --wait-for-expect-header string          Expect response header pattern.
      --wait-for-expect-status-code int        Expect response code e.g. 200, 204, ... .
      --wait-for-insecure-skip-tls-verify      Skips tls certificate checks for the HTTPS request.
      --wait-for-interval duration             Interval time between each loop. (default 1s)
      --wait-for-no-redirect                   Do not follow HTTP 3xx redirects.
      --wait-for-timeout duration              Timeout is the maximum amount of time that Wait4X will wait for a checking operation, 0 is unlimited. (default 10s)
```

Fixes #142.

### CODE REVIEW

**Comment by fzipi** [comment]
Initial console tests look good:
- using status code 200:
```
./go-ftw run --wait-for-host http://localhost:80/status/200 --wait-for-expect-status-code 200 -d tests/regression
10:17AM INF [HTTP] Checking the http://localhost:80/status/200 ... v=0
🛠️  Starting tests!
🚀 Running go-ftw!
👉 executing tests in file 911100.yaml
	running 911100-1: ✔ passed in 65.929306ms (RTT 83.769132ms)
	running 911100-2: ✔ passed in 20.632873ms (RTT 55.852238ms)
...
```
Failed:
```
./go-ftw/go-ftw run --wait-for-host http://localhost:80/status/200 --wait-for-expect-status-code 123 -d tests/regression
10:25AM INF [HTTP] Checking the http://localhost:80/status/200 ... v=0
10:25AM ERR Expectation failed error="the status code doesn't expect" actual=200 expect=123
10:25AM INF [HTTP] Checking the http://localhost:80/status/200 ... v=0
10:25AM ERR Expectation failed error="the status code doesn't expect" actual=200 expect=123
10:25AM INF [HTTP] Checking the http://localhost:80/status/200 ... v=0
10:25AM ERR Expectation failed error="the status code doesn't expect" actual=200 expect=123
...
```
- using json response syntax
```
/go-ftw run --wait-for-host http://localhost:80/get --wait-for-expect-body-json "headers.User-Agent" -d tests/regression
10:23AM INF [HTTP] Checking the http://localhost:80/get ... v=0
🛠️  Starting tests!
🚀 Running go-ftw!
👉 executing tests in file 911100.yaml
	running 911100-1: ✔ passed in 18.980143ms (RTT 58.225661ms)
	running 911100-2: ✔ passed in 18.594237ms (RTT 55.91404ms)
	running 911100-3: ✔ passed in 17.570498ms (RTT 55.772243ms)
```
Failed response:
```
./go-ftw run --wait-for-host http://localhost:80/get --wait-for-expect-body-json "headeers.user_agent" -d tests/regression
10:23AM INF [HTTP] Checking the http://localhost:80/get ... v=0
10:23AM ERR Expectation failed error="the JSON doesn't match" actual="{\n  \"args\": {}, \n  \"headers\": {\n    \"Connection..." expect=headeers.user_agent
10:23AM INF [HTTP] Checking the http://localhost:80/get ... v=0
10:23AM ERR Expectation failed error="the JSON doesn't match" actual="{\n  \"args\": {}, \n  \"headers\": {\n    \"Connection..." expect=headeers.user_agent
```

**Comment by jcchavezs** [comment]
Thanks @fzipi. I wonder about the use cases for the other flags.

File: cmd/run.go [hunk 1]
```diff
@@ -1,105 +1,171 @@
...
 	runCmd.Flags().StringP("output", "o", "normal", "output type for ftw tests. \"normal\" is the default.")
+	runCmd.Flags().StringP("file", "f", "", "output file for ftw tests. \"os.Stdout\" is the default.")
 	runCmd.Flags().BoolP("time", "t", false, "show time spent per test")
 	runCmd.Flags().BoolP("show-failures-only", "", false, "shows only the results of failed tests")
 	runCmd.Flags().Duration("connect-timeout", 3*time.Second, "timeout for connecting to endpoints during test execution")
 	runCmd.Flags().Duration("read-timeout", 1*time.Second, "timeout for receiving responses during test execution")
 	runCmd.Flags().Int("max-marker-retries", 20, "maximum number of times the search for log markers will be repeated.\nEach time an additional request is sent to the web server, eventually forcing the log to be flushed")
 	runCmd.Flags().Int("max-marker-log-lines", 500, "maximum number of lines to search for a marker before aborting")
+	runCmd.Flags().String("wait-for-host", "", "Wait for host to be available before running tests.")
+	runCmd.Flags().Duration("wait-for-interval", 1*time.Second, "Interval time between each loop.")
```
Comment (theseion) [inline cmd/run.go:45]:
```suggestion
	runCmd.Flags().Duration("wait-delay", 1*time.Second, "Time to wait between retries for all wait operations.")
```

File: cmd/run_test.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  description: "Test file for go-ftw"
+tests:
+  - # Standard GET request
+    test_title: 1234
+    stages:
+      - stage:
+          input:
+            dest_addr: "127.0.0.1"
+            method: "GET"
+            port: {{ .Port }}
```
Comment (theseion) [inline cmd/run_test.go:34]: Of course! That's much better than using regular expressions to replace stuff... We should fix that in all of the test YAML's we use.

**Review by theseion — CHANGES_REQUESTED** [review]

File: cmd/run_test.go (hunk 2 above)
Comment (fzipi) [reply cmd/run_test.go:34]: I'm planning to overhaul all the testing stuff. This is part of the change ;)

**Review by theseion — APPROVED** [review]

File: cmd/run.go [hunk 3]
```diff
@@ -1,105 +1,171 @@
...
+		if waiterErr != nil {
+			return waiterErr
+		}
+	}
+	// use outputFile to write to file
+	var outputFile *os.File
+	if outputFilename == "" {
+		outputFile = os.Stdout
+	} else {
+		outputFile, err = os.Open(outputFilename)
```
Comment (jcchavezs) [inline cmd/run.go:147]: We should close the file at the end.

File: ftwhttp/client_test.go [hunk 4]
```diff
@@ -50,7 +50,7 @@ func TestDoRequest(t *testing.T) {
 
 func TestGetTrackedTime(t *testing.T) {
 	d := &Destination{
-		DestAddr: "httpbin.org",
+		DestAddr: "httpbingo.org",
```
Comment (jcchavezs) [inline ftwhttp/client_test.go:53]: At some point we should stop depending on external deps.

Comment (fzipi) [reply ftwhttp/client_test.go:53]: Agreed. I'm overhauling all tests and will deal with this in a future PR.

File: cmd/run.go (hunk 3 above)
Comment (fzipi) [reply cmd/run.go:147]: Good catch.

**Pull request merged**